Enh/accelerated asset refactor

dizcza opened the pull request:
* minor simplification of the code without changing the logic
* docstring
* naming convention

dizcza committed:
simplified

dizcza committed:
asset documentation without renaming

dizcza committed:
asset renaming1

dizcza committed:
asset fixed tests

JuliaSprenger commented on elephant/asset.py:
```diff
@@ -35,7 +35,7 @@
 
 5) Extract sequences of synchronous events associated to each worm
```
I am surprised the term 'worm' is still showing up here. I don't think this is introduced anywhere here, so maybe we should replace it by  sequence of synchronous event / asset depending on the context?

dizcza replied:
I remember my confusion in my first month when I looked in asset and one of the first lines of examples says (it still does now!):

```Extract sequences of synchronous events associated to each worm```

I mean, it's not in uppercase :)

JuliaSprenger commented on elephant/asset.py:
```diff
@@ -376,8 +364,9 @@ def intersection_matrix(
-    spikes_per_bin_x = bsts_x.sum(axis=0)
-    spikes_per_bin_y = bsts_y.sum(axis=0)
+    # 'A1' property returns self as a flattened ndarray.
+    spikes_per_bin_x = bsts_x.sum(axis=0).A1
```
maybe `ravel` or `flatten` would be more telling than `A1`? I haven't seen that being used anywhere else.

dizcza replied:
why not learn new things? :)
1. your approach will be lengthy - `np.array(bsts_x.sum(axis=0)).squeeze()`, - because 'flatten()' of a numpy matrix is still a numpy matrix. And a numpy matrix is always 2d (I want 1d).
2. `.A1` property shares the data with the original matrix! No copy is involved.

JuliaSprenger replied:
Ok, I see the point. Fine with me to leave it as it is.

JuliaSprenger commented:
I see why you replaced `sse` by `synchronous_events` and not by `sequence_of_synchronous_events`. However, I think this might lead to confusion in the future, so this convention should be introduced and highlighted in the beginning of the documentation.
Or does anybody have a better idea on how to extend `sse` to make it more readable, but not too long?

kohlerca commented on elephant/asset.py:
```diff
@@ -158,19 +158,19 @@ def _transactions(spiketrains, binsize, t_start=None, t_stop=None, ids=None):
         A list of `neo.SpikeTrain` objects, or list of pairs
         (Train_ID, `neo.SpikeTrain`), where Train_ID can be any hashable object
-    binsize: pq.Quantity
+    binsize : pq.Quantity
```
This should be `bin_size`

dizcza replied:
right

dizcza committed:
renamed private func _interpolate_signals -> _interpolate_firing_rates; supplementary notebook link

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -457,37 +442,38 @@ def mask_matrices(matrices, thresholds):
     r"""
-    Given a list of points on the real plane, identified by their absciss x
-    and ordinate y, compute a stretched transformation of the Euclidean
+    Given a list of points on the real plane, identified by their absciss `x`
```
Shouldn't it be "abscissa" not "absciss"? At least accoring to a quick check in a dictionary.

dizcza replied:
I'll rename it to abscissa

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -457,37 +442,38 @@ def mask_matrices(matrices, thresholds):
     ----------
-    x : numpy.ndarray
+    x : np.ndarray
         array of abscissa of all points among which to compute the distance
```
And here "abscissas" or "abscissae". I did not check whether the same should be corrected in other locations in the code.

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -551,44 +537,45 @@ def cluster_matrix_entries(mat, eps=10, min_neighbors=2, stretch=5):
     -------
-    cmat : numpy.ndarray of integers
+    cmat : np.ndarray
         a matrix with the same shape of mat, each of whose elements is either
```
The same shape as `mask_matrix`, not mat.

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -842,63 +820,55 @@ def probability_matrix_analytical(
     Parameters
     ----------
     spiketrains : list of neo.SpikeTrains
```
neo.SpikeTrain (without s)

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -1036,26 +1006,26 @@ def _jsf_uniform_orderstat_3d(u, alpha, n, verbose=False):
-    u : numpy.ndarray of shape (A, d)
+    Parameters
+    ----------
+    u : (A,d) np.ndarray
```
Why is the first dimension called "A"? It is a bit confusing, because the capital letter makes the impression, that it refers to sth specific, but no "A" is mentioned across the whole docstring. 
Also, in the description above the are no ticks around math-related expressions, but this is because the function will not be visible on the docs website, right?

dizcza replied:
Yes, I focused on public functions and left most of the private ones untouched.
I might jump over some private functions in a rush before realizing that they won't be shown anyway.

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -1170,22 +1138,22 @@ def _pmat_neighbors(mat, filter_shape, nr_largest):
     -------
-    L : ndarray
+    L : np.ndarray
         a matrix of shape (nr_largest, l, w) containing along the first
```
n_largest (not nr_largest)

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -1341,38 +1310,41 @@ def extract_sse(spiketrains, binsize, cmat, spiketrains_y=None,
+        The binsize used in the computation of the cluster matrix `cmat`.
+    cmat: (n,n) np.ndarray
+        The matrix of shape (n, n), where `n` is the length of `x_edges` and
+        `y_edges`, representing the cluster matrix in worms analysis
```
Another worm here. Should they be replaced by other expression as Julia suggested?

dizcza replied:
I'll leave worms in comments and variable names

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -1424,33 +1396,32 @@ def extract_sse(spiketrains, binsize, cmat, spiketrains_y=None,
+    sse1, sse2 : dict
+        each is a dictionary of pixel positions (i, j) as keys and sets S of
         synchronous events as values (see above).
     intersection : str, optional
```
"{'pixelwise', 'linkwise'}" instead of "str, optional"

PaulinaDabrowska replied:
And than the second sentence in the description can be removed.

PaulinaDabrowska commented on elephant/asset.py:
```diff
@@ -1485,47 +1461,50 @@ def sse_intersection(sse1, sse2, intersection='linkwise'):
+    sse1, sse2 : dict
+        Dictionaries of pixel positions (i, j) as keys and sets S of
+        synchronous events as values.
     difference : str, optional
```
The same as above should be {'pixelwise', 'linkwise'}.

PaulinaDabrowska reviewed (commented):
I focused mostly on the naming, so most of my comments look very picky - sorry for this. In general, well done, that was a lot of work.

Kleinjohann commented:
> I see why you replaced `sse` by `synchronous_events` and not by `sequence_of_synchronous_events`. However, I think this might lead to confusion in the future, so this convention should be introduced and highlighted in the beginning of the documentation.
> Or does anybody have a better idea on how to extend `sse` to make it more readable, but not too long?

I agree, this could lead to problems. Also, if we do stick with this new name we should rename all of the `synchronous_events_is_*` functions to be at least a bit closer to grammatical correctness.

kohlerca committed:
General revision of docstrings in asset.py

kohlerca commented on elephant/asset.py:
```diff
@@ -267,70 +267,58 @@ def intersection_matrix(
-        If None (default) the attribute t_stop of the SpikeTrains is used
-        (if the same for all spike trains).
-        Default: None
     norm : int, optional
```
This could be changed to `normalization`.
Also, 0 or None have the same effect.
One option would be to use strings and give the meaning of each normalization operation.

dizcza replied:
@stellalessandra @Kleinjohann what do you think?

Kleinjohann replied:
No strong opinion. If we can come up with short descriptive strings that would be good. `None` should still work and be the default though.

kohlerca commented on elephant/asset.py:
```diff
@@ -267,70 +267,58 @@ def intersection_matrix(
+    imat : (n,n) np.ndarray
+        The intersection matrix (of type float) of a list of spike trains. Has
+        the shape (n,n), where `n` is the number of bins time was discretized in.
+    x_edges : (n,n) np.ndarray
```
Inside the function, `x_edges` and `y_edges` are named `xx` and `yy` respectively. According to our conventions, we should use the same name as in the code, but I guess refactoring the names inside the function would add to clarity.

kohlerca commented on elephant/asset.py:
```diff
@@ -457,37 +442,38 @@ def mask_matrices(matrices, thresholds):
     stretch : float
         maximum stretching factor, applied if the line connecting the points
         has inclination 90 + ref_angle
     ref_angle : float
```
Is `ref_angle` in degrees? Then should specify in the description.

dizcza replied:
yes
moreover, it should be renamed to `reference_angle`
but I didn't look into private functions

kohlerca commented on elephant/asset.py:
```diff
@@ -1233,56 +1201,59 @@ def _pmat_neighbors(mat, filter_shape, nr_largest):
 
 def joint_probability_matrix(
-        pmat, filter_shape, nr_largest=None, alpha=0, pvmin=1e-5,
+        pmat, filter_shape, n_largest=None, alpha=0, p_value_min=1e-5,
```
Parameter name should be `min_p_value` according to the naming conventions.

dizcza replied:
right

kohlerca commented on elephant/asset.py:
```diff
@@ -612,101 +599,93 @@ def cluster_matrix_entries(mat, eps=10, min_neighbors=2, stretch=5):
         spiketrains, binsize, spiketrains_y=None,
         t_start_x=None, t_start_y=None, t_stop_x=None, t_stop_y=None,
-        surr_method='dither_spike_train', j=None, n_surr=100, verbose=False):
+        surr_method='dither_spike_train', jitter=None, n_surr=100,
```
Parameter named `surrogate_method` is clearer.
As in the other functions, `binsize` should be `bin_size`.
And `n_surr` should be `n_surrogates`.

dizcza replied:
I agree with all

Kleinjohann committed:
Account for Rounding Errors when Binning Spike Trains (#305)

* Implement fix for rounding issue in binning

* Make tolerance a parameter, add warnings

dizcza commented:
> > I see why you replaced `sse` by `synchronous_events` and not by `sequence_of_synchronous_events`. However, I think this might lead to confusion in the future, so this convention should be introduced and highlighted in the beginning of the documentation.
> > Or does anybody have a better idea on how to extend `sse` to make it more readable, but not too long?
> 
> I agree, this could lead to problems. Also, if we do stick with this new name we should rename all of the `synchronous_events_is_*` functions to be at least a bit closer to grammatical correctness.

Do you mean renaming `is` -> `all` and `contains` -> `contain`?

I looked into this issue of is/are in programming languages quite some time ago. There is no convention, and developers still use `is` instead of grammatically correct `are`.
Anyway, I don't have any preference here except that we should not call these functions as `sse_xxx`.

@Kleinjohann will you rename them as you wish and update the docs accordingly? Because I don't understand what problems could arise and how exactly we should highlight this at the beginning of asset docs.

Commit:
Second general revision of docstrings in asset.py

kohlerca commented on elephant/asset.py:
```diff
@@ -1128,34 +1159,43 @@ def _jsf_uniform_orderstat_3d(u, alpha, n, verbose=False):
+        
+    Raises
+    ------
+    AssertionError
```
Could replace the `assert` to raise ValueError, since assertions were not supposed to be used in Elephant

pbouss committed:
added uniform dithering method with refr period (#297)

* upd surrogate type

* Removed dither_spikes_with_refr_period from surrogates wrapper

Co-authored-by: p-bouss <peter.bouss@googlemail.com>
Co-authored-by: dizcza <dizcza@gmail.com>

dizcza committed:
Merge branch 'master' into enh/accelerated_asset_refactor

dizcza committed:
asset example (will be removed)

dizcza committed:
asset OOP

dizcza committed:
adapted asset test to follow OOP

dizcza committed:
jmat filter_shape=(5,1) by default; simple example

dizcza commented:
ASSET in OOP

**Fixed bugs**
1. previously, intersection_matrix `norm` parameter is always None by default (there were no ways to change it since probability_matrix_xxx didn't have `norm` param)
2. previously, `extract_synchronous_events` ignored t_start and t_stop, that were passed in the intermediate steps of asset analysis, and always computed them from the input spike trains.

**Changed behavior**
1. Every '_matrix' function takes another matrix as input (None by default) and outputs exactly one matrix.
2. probability_matrix_xxx functions take `imat` as input (None by default)
3. `intersection_matrix` and both probability_matrix_xxx functions don't return x_edges and y_edges. Now they are properties of ASSET class.
4. verbose is True by default

**Not sure**
1. Does extract_sychronous_events work with any `spiketrains` or only those that has been used in the very input?
2. There are 2 functions - `mask_matrices` and `cluster_matrix_entries` - that can operate without a notion of a spiketrain. In tests, now I force creating an ASSET object with spiketrains to call these function. If you do call these functions without doing and presaving the intermediate steps (although for me it's hard to image, but let's suppose), we can easily move the logic of these functions outside of the ASSET class and still do the following in the class:
```python
if mask_matrix is None
	mask_matrix = self.mask_matrices()
return cluster_matrix_entries(mask_matrix, ...)  # cluster_matrix_entries is defined outside
```
and the same for `mask_matrices()` function.
3. `joint_probability_matrix` requires filter_shape. I made it (5,1) by default. We could of course make it positional (no default value) and force users to think what the valid range of `filter_shape` is. Making it positional will also break the nice structure that I provided and the minimal knowledge of asset. We can highlight this caveat in the doc and show in the tutorial.

**Simple API**
Check out the example https://github.com/INM-6/elephant/blob/enh/accelerated_asset_refactor/elephant/asset_example.py
I'll remove it before merging and put in the Example docstring alongside with the long way (old style).

**What is left**
1. [x] pep8
4. [x] Group the functions in the module documentation as it's done in kernels.py and statistics.py.
5. [x] Move the simple example above to docs.
6. [x] Build the docs and see that everything is fine.
7. Merge the PR.

**Won't do**
2. Update the documentation (done) and tutorial (not done), describing why and _how_ the user has to change the important parameters - `surrogate_dt` in `probability_matrix_montecarlo` and `filter_shape` in `joint_probability_matrix`.
3. Once updated, move the tutorial from https://github.com/INM-6/elephant-tutorials/blob/master/asset_showcase.ipynb to Elephant (data has to be uploaded to gin).

dizcza committed:
added utf-8 encoding header

dizcza committed:
intersection_matrix() does not take spiketrains parameters

dizcza committed:
asset documentation is done

dizcza committed:
normalization parameter meaningful name

dizcza committed:
deleted asset_example

Kleinjohann commented:
> **Not sure**
> 
>     1. Does extract_sychronous_events work with any `spiketrains` or only those that has been used in the very input?
> 
>     2. There are 2 functions - `mask_matrices` and `cluster_matrix_entries` - that can operate without a notion of a spiketrain. In tests, now I force creating an ASSET object with spiketrains to call these function. If you do call these functions without doing and presaving the intermediate steps (although for me it's hard to image, but let's suppose), we can easily move the logic of these functions outside of the ASSET class and still do the following in the class:

1. it should only ever be used with those in the input, nothing else really makes any sense
2. Being able to call these separately might make a workflow more understandable since it can then mirror the execution order of ASSET functions. However, it might be worth it to sacrifice that to be able to put all of the code together for better maintainability.

Kleinjohann commented:
> **Fixed bugs**
> 
>     1. previously, intersection_matrix `norm` parameter is always None by default (there were no ways to change it since probability_matrix_xxx didn't have `norm` param)

This was not a bug, `intersection_matrix` is only supposed to be called with `norm=None` from the `probability_matrix_*` functions. In the Monte-Carlo case this might work, but only if you always use the same normalisation, also for the surrogates. In the analytical version the Poisson estimation uses a normalisation that only matches `norm=None` in `intersection_matrix`.

Kleinjohann commented:
> **Not sure**
> 
> [...]
>
> 3. `joint_probability_matrix` requires filter_shape. I made it (5,1) by default. We could of course make it positional (no default value) and force users to think what the valid range of `filter_shape` is. Making it positional will also break the nice structure that I provided and the minimal knowledge of asset. We can highlight this caveat in the doc and show in the tutorial.

Sounds good to me, as long as all defaults make some sort of sense together, e.g. work in the simple test (which they seemed to do in the file you now removed again, right?)

dizcza commented:
> > **Not sure**
> > [...]
> > 
> > 1. `joint_probability_matrix` requires filter_shape. I made it (5,1) by default. We could of course make it positional (no default value) and force users to think what the valid range of `filter_shape` is. Making it positional will also break the nice structure that I provided and the minimal knowledge of asset. We can highlight this caveat in the doc and show in the tutorial.
> 
> Sounds good to me, as long as all defaults make some sort of sense together, e.g. work in the simple test (which they seemed to do in the file you now removed again, right?)

filter_shape=(5,1) is taken from the tutorial (asset_showcase). It works for my test (even smaller shape might work as well) - I moved the test in the Examples section of the module docstring.

dizcza commented:
> Being able to call these separately might make a workflow more understandable since it can then mirror the execution order of ASSET functions. However, it might be worth it to sacrifice that to be able to put all of the code together for better maintainability.

I leave it as it is now then. After you use it for a while, you get the final impression and can change the behavior as you wish.

kohlerca commented on doc/bib/elephant.bib:
```diff
@@ -62,4 +62,14 @@ @article{Riehle97_1950
+  journal={PLoS computational biology},
+  volume={12},
+  number={7},
+  doi={10.1371/journal.pcbi.1004939}
```
Missing the page: e1004939

dizcza replied:
I'm not very comfortable with citing
should it be `pages = {e1004939}`?

kohlerca replied:
Yes. Although they do not have actual pages in PLoS, `pages = {e1004939}` corresponds to a record with only the starting page field in other formats (such RIS).

dizcza committed:
unwrapped asset chain call

dizcza committed:
added pages in asset bibtex

Commit:
pep8

Commit:
Remove unnecessary parameter alpha

Commit:
Make thresholds argument positional

Commit:
Make filter_shape argument positional

Commit:
Make cluster_matrix_entries arguments positional

Commit:
Make n_largest argument positional

stellalessandra committed:
modified example in documentation

stellalessandra committed:
Made n_surrogates argument positional

dizcza commented on elephant/asset.py:
```diff
@@ -444,12 +443,13 @@ def _jsf_uniform_orderstat_3d(u, alpha, n, verbose=False):
     # Compute the log of the integral's coefficient
-    logK = np.sum(np.log(np.arange(1, n + 1))) - n * np.log(1 - alpha)
-    # Add to the 3D matrix u a bottom layer equal to alpha and a
+    logK = np.sum(np.log(np.arange(1, n + 1))) - n * log_1
```
log(1) is zero
if it's not a mistake, drop a comment, explaining why the log power simplifies to `1`. Maybe, because it should be `1-alpha` and `alpha` is zero?
` - n*0` is not necessary then, up to you to leave it

Kleinjohann replied:
Yes, it is correct, we just set `alpha=0` everywhere now. I kept it for the same reason we have the other `log(1)` in the same function; to try and make it more clear how the formula was derived. But I guess you just proved that it only raises more questions here, so I removed it. The formula with `alpha=0` will probably be more familiar to potential readers anyway.

dizcza replied:
Is there a specific formula when `alpha=0` in the paper? If yes, then the comment of not subtracting `n log(1-alpha)` is not needed.

dizcza commented on elephant/asset.py:
```diff
@@ -1539,17 +1537,14 @@ def joint_probability_matrix(self, pmat, filter_shape=(5, 1),
+        n_largest : int
             The number of the largest neighbors to collect for each entry in
             `jmat`. If None, the filter length `l` of `filter_shape` is used.
             Default: None.
```
remove default and the sentence "if None, ..."

dizcza commented on elephant/asset.py:
```diff
@@ -1692,20 +1687,17 @@ def cluster_matrix_entries(mask_matrix, eps=10, min_neighbors=2,
             The boolean matrix, whose elements with positive values are to be
             clustered. The output of :func:`ASSET.mask_matrices`.
-        eps : float, optional
+        eps : float
```
just spot it
let's rename it
how about `max_distance`?

Commit:
Clean up vanishing contribution to logK

Commit:
Clean up documentation of n_largest default

Commit:
Rename eps to max_distance

Commit:
pep8 in test_asset

dizcza commented on elephant/asset.py:
```diff
@@ -1687,7 +1687,7 @@ def cluster_matrix_entries(mask_matrix, eps, min_neighbors, stretch):
             The boolean matrix, whose elements with positive values are to be
             clustered. The output of :func:`ASSET.mask_matrices`.
-        eps : float
+        max_distance : float
```
if the max distance is denoted as :math:`\eps` in the paper, it's always a good practice to relate the argument/variable names to the math names in the paper
but this is a minor thing and even if it's the case in asset, I'll not pursue it here.

dizcza committed:
cleanup docs for n_largest cannot be None anymore; removed if pmat is None statement in jmat
